tests/test_pytest_plugin.py 4 failures when pytest-asyncio is installed

pytest-asyncio seems to interfere with anyio's pytest plugin somehow:

```
tests/test_pytest_plugin.py FF..FF
```

You can easily reproduce by hacking in pytest-asyncio dep into the tox env, i.e.:
```
[testenv]
deps =
    pytest-asyncio
```

Full output follows.

```pytb
====================================================== FAILURES ======================================================
____________________________________________________ test_plugin _____________________________________________________
/tmp/anyio-3.2.1/tests/test_pytest_plugin.py:65: in test_plugin
    result.assert_outcomes(passed=3 * len(get_all_backends()), skipped=len(get_all_backends()))
E   AssertionError: assert {'errors': 0,...pped': 2, ...} == {'errors': 0,...pped': 2, ...}
E     Omitting 4 identical items, use -vv to show
E     Differing items:
E     {'passed': 5} != {'passed': 6}
E     {'failed': 1} != {'failed': 0}
E     Use -v to get the full diff
------------------------------------------------ Captured stdout call ------------------------------------------------
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /tmp/anyio-3.2.1/.tox/py38/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/tmp/anyio-3.2.1/.hypothesis/examples')
rootdir: /tmp/pytest-of-mgorny/pytest-5/test_plugin0
plugins: anyio-3.2.1, hypothesis-6.14.0, mock-3.6.1, asyncio-0.15.1
collecting ... collected 8 items

test_plugin.py::test_marked_test[asyncio] PASSED                         [ 12%]
test_plugin.py::test_marked_test[trio] PASSED                            [ 25%]
test_plugin.py::test_async_fixture_from_marked_test[asyncio] PASSED      [ 37%]
test_plugin.py::test_async_fixture_from_marked_test[trio] PASSED         [ 50%]
test_plugin.py::test_async_fixture_from_sync_test[asyncio] PASSED        [ 62%]
test_plugin.py::test_async_fixture_from_sync_test[trio] FAILED           [ 75%]
test_plugin.py::test_skip_inline[asyncio] SKIPPED (Test that skippin...) [ 87%]
test_plugin.py::test_skip_inline[trio] SKIPPED (Test that skipping w...) [100%]

=================================== FAILURES ===================================
___________________ test_async_fixture_from_sync_test[trio] ____________________

anyio_backend_name = 'trio', async_fixture = 'asyncio'

    def test_async_fixture_from_sync_test(anyio_backend_name, async_fixture):
        # Test that regular functions can use async fixtures too
>       assert async_fixture == anyio_backend_name
E       AssertionError: assert 'asyncio' == 'trio'
E         - trio
E         + asyncio

test_plugin.py:22: AssertionError
=========================== short test summary info ============================
FAILED test_plugin.py::test_async_fixture_from_sync_test[trio] - AssertionErr...
==================== 1 failed, 5 passed, 2 skipped in 0.09s ====================
____________________________________________________ test_asyncio ____________________________________________________
/tmp/anyio-3.2.1/tests/test_pytest_plugin.py:138: in test_asyncio
    result.assert_outcomes(passed=2, failed=1, errors=2)
E   AssertionError: assert {'errors': 1,...pped': 0, ...} == {'errors': 2,...pped': 0, ...}
E     Omitting 5 identical items, use -vv to show
E     Differing items:
E     {'errors': 1} != {'errors': 2}
E     Use -v to get the full diff
------------------------------------------------ Captured stdout call ------------------------------------------------
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /tmp/anyio-3.2.1/.tox/py38/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/tmp/anyio-3.2.1/.hypothesis/examples')
rootdir: /tmp/pytest-of-mgorny/pytest-5/test_asyncio0
plugins: anyio-3.2.1, hypothesis-6.14.0, mock-3.6.1, asyncio-0.15.1
collecting ... collected 4 items

test_asyncio.py::TestClassFixtures::test_class_fixture_in_test_method ERROR [ 25%]
test_asyncio.py::test_callback_exception_during_test FAILED              [ 50%]
test_asyncio.py::test_callback_exception_during_setup PASSED             [ 75%]
test_asyncio.py::test_callback_exception_during_teardown PASSED          [100%]

==================================== ERRORS ====================================
____ ERROR at setup of TestClassFixtures.test_class_fixture_in_test_method _____
ScopeMismatch: You tried to access the 'function' scoped fixture 'event_loop' with a 'class' scoped request object, involved factories
../../../anyio-3.2.1/.tox/py38/lib/python3.8/site-packages/pytest_asyncio/plugin.py:151:  def wrapper(*args, **kwargs)
../../../anyio-3.2.1/.tox/py38/lib/python3.8/site-packages/pytest_asyncio/plugin.py:224:  def event_loop(request)
=================================== FAILURES ===================================
_____________________ test_callback_exception_during_test ______________________

    def callback():
        nonlocal started
        started = True
>       raise Exception('foo')
E       Exception: foo

test_asyncio.py:22: Exception
=========================== short test summary info ============================
FAILED test_asyncio.py::test_callback_exception_during_test - Exception: foo
ERROR test_asyncio.py::TestClassFixtures::test_class_fixture_in_test_method
===================== 1 failed, 2 passed, 1 error in 0.04s =====================
------------------------------------------------- Captured log call --------------------------------------------------
ERROR    asyncio:plugin.py:142 Exception in callback <function setup_fail_fixture.<locals>.callback at 0x7ff339de9e50>
handle: <Handle setup_fail_fixture.<locals>.callback>
Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 61, in uvloop.loop.Handle._run
  File "/tmp/pytest-of-mgorny/pytest-5/test_asyncio0/conftest.py", line 12, in callback
    raise RuntimeError('failing fixture setup')
RuntimeError: failing fixture setup
ERROR    asyncio:base_events.py:1707 Exception in callback teardown_fail_fixture.<locals>.callback() at /tmp/pytest-of-mgorny/pytest-5/test_asyncio0/conftest.py:20
handle: <Handle teardown_fail_fixture.<locals>.callback() at /tmp/pytest-of-mgorny/pytest-5/test_asyncio0/conftest.py:20>
Traceback (most recent call last):
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/tmp/pytest-of-mgorny/pytest-5/test_asyncio0/conftest.py", line 21, in callback
    raise RuntimeError('failing fixture teardown')
RuntimeError: failing fixture teardown
____________________________________________ test_hypothesis_module_mark _____________________________________________
/tmp/anyio-3.2.1/tests/test_pytest_plugin.py:233: in test_hypothesis_module_mark
    result.assert_outcomes(passed=len(get_all_backends()) + 1, xfailed=len(get_all_backends()))
E   AssertionError: assert {'errors': 0,...pped': 0, ...} == {'errors': 0,...pped': 0, ...}
E     Omitting 4 identical items, use -vv to show
E     Differing items:
E     {'passed': 2} != {'passed': 3}
E     {'failed': 1} != {'failed': 0}
E     Use -v to get the full diff
------------------------------------------------ Captured stdout call ------------------------------------------------
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /tmp/anyio-3.2.1/.tox/py38/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/tmp/anyio-3.2.1/.hypothesis/examples')
rootdir: /tmp/pytest-of-mgorny/pytest-5/test_hypothesis_module_mark0
plugins: anyio-3.2.1, hypothesis-6.14.0, mock-3.6.1, asyncio-0.15.1
collecting ... collected 5 items

test_hypothesis_module_mark.py::test_hypothesis_wrapper[asyncio] PASSED  [ 20%]
test_hypothesis_module_mark.py::test_hypothesis_wrapper[trio] FAILED     [ 40%]
test_hypothesis_module_mark.py::test_hypothesis_wrapper_regular PASSED   [ 60%]
test_hypothesis_module_mark.py::test_hypothesis_wrapper_failing[asyncio] XFAIL [ 80%]
test_hypothesis_module_mark.py::test_hypothesis_wrapper_failing[trio] XFAIL [100%]

=================================== FAILURES ===================================
________________________ test_hypothesis_wrapper[trio] _________________________

    @given(x=just(1))
>   async def test_hypothesis_wrapper(x):

test_hypothesis_module_mark.py:9: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/tmp/anyio-3.2.1/.tox/py38/lib/python3.8/site-packages/pytest_asyncio/plugin.py:193: in inner
    task = asyncio.ensure_future(coro, loop=_loop)
/usr/lib/python3.8/asyncio/tasks.py:672: in ensure_future
    task = loop.create_task(coro_or_future)
/usr/lib/python3.8/asyncio/base_events.py:429: in create_task
    self._check_closed()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_UnixSelectorEventLoop running=False closed=True debug=False>

    def _check_closed(self):
        if self._closed:
>           raise RuntimeError('Event loop is closed')
E           RuntimeError: Event loop is closed

/usr/lib/python3.8/asyncio/base_events.py:508: RuntimeError
---------------------------------- Hypothesis ----------------------------------
Falsifying example: test_hypothesis_wrapper(
    x=1,
)
=========================== short test summary info ============================
FAILED test_hypothesis_module_mark.py::test_hypothesis_wrapper[trio] - Runtim...
==================== 1 failed, 2 passed, 2 xfailed in 0.35s ====================
___________________________________________ test_hypothesis_function_mark ____________________________________________
/tmp/anyio-3.2.1/tests/test_pytest_plugin.py:272: in test_hypothesis_function_mark
    result.assert_outcomes(passed=2 * len(get_all_backends()), xfailed=2 * len(get_all_backends()))
E   AssertionError: assert {'errors': 0,...pped': 0, ...} == {'errors': 0,...pped': 0, ...}
E     Omitting 4 identical items, use -vv to show
E     Differing items:
E     {'passed': 2} != {'passed': 4}
E     {'failed': 2} != {'failed': 0}
E     Use -v to get the full diff
------------------------------------------------ Captured stdout call ------------------------------------------------
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /tmp/anyio-3.2.1/.tox/py38/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/tmp/anyio-3.2.1/.hypothesis/examples')
rootdir: /tmp/pytest-of-mgorny/pytest-5/test_hypothesis_function_mark0
plugins: anyio-3.2.1, hypothesis-6.14.0, mock-3.6.1, asyncio-0.15.1
collecting ... collected 8 items

test_hypothesis_function_mark.py::test_anyio_mark_first[asyncio] PASSED  [ 12%]
test_hypothesis_function_mark.py::test_anyio_mark_first[trio] FAILED     [ 25%]
test_hypothesis_function_mark.py::test_anyio_mark_last[asyncio] PASSED   [ 37%]
test_hypothesis_function_mark.py::test_anyio_mark_last[trio] FAILED      [ 50%]
test_hypothesis_function_mark.py::test_anyio_mark_first_fail[asyncio] XFAIL [ 62%]
test_hypothesis_function_mark.py::test_anyio_mark_first_fail[trio] XFAIL [ 75%]
test_hypothesis_function_mark.py::test_anyio_mark_last_fail[asyncio] XFAIL [ 87%]
test_hypothesis_function_mark.py::test_anyio_mark_last_fail[trio] XFAIL  [100%]

=================================== FAILURES ===================================
_________________________ test_anyio_mark_first[trio] __________________________

    @pytest.mark.anyio
>   @given(x=just(1))

test_hypothesis_function_mark.py:7: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/tmp/anyio-3.2.1/.tox/py38/lib/python3.8/site-packages/pytest_asyncio/plugin.py:193: in inner
    task = asyncio.ensure_future(coro, loop=_loop)
/usr/lib/python3.8/asyncio/tasks.py:672: in ensure_future
    task = loop.create_task(coro_or_future)
/usr/lib/python3.8/asyncio/base_events.py:429: in create_task
    self._check_closed()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_UnixSelectorEventLoop running=False closed=True debug=False>

    def _check_closed(self):
        if self._closed:
>           raise RuntimeError('Event loop is closed')
E           RuntimeError: Event loop is closed

/usr/lib/python3.8/asyncio/base_events.py:508: RuntimeError
---------------------------------- Hypothesis ----------------------------------
Falsifying example: test_anyio_mark_first(
    x=1,
)
__________________________ test_anyio_mark_last[trio] __________________________

>   ???

test_hypothesis_function_mark.py:13: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/tmp/anyio-3.2.1/.tox/py38/lib/python3.8/site-packages/pytest_asyncio/plugin.py:193: in inner
    task = asyncio.ensure_future(coro, loop=_loop)
/usr/lib/python3.8/asyncio/tasks.py:672: in ensure_future
    task = loop.create_task(coro_or_future)
/usr/lib/python3.8/asyncio/base_events.py:429: in create_task
    self._check_closed()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_UnixSelectorEventLoop running=False closed=True debug=False>

    def _check_closed(self):
        if self._closed:
>           raise RuntimeError('Event loop is closed')
E           RuntimeError: Event loop is closed

/usr/lib/python3.8/asyncio/base_events.py:508: RuntimeError
---------------------------------- Hypothesis ----------------------------------
Falsifying example: test_anyio_mark_last(
    x=1,
)
=========================== short test summary info ============================
FAILED test_hypothesis_function_mark.py::test_anyio_mark_first[trio] - Runtim...
FAILED test_hypothesis_function_mark.py::test_anyio_mark_last[trio] - Runtime...
==================== 2 failed, 2 passed, 4 xfailed in 0.70s ====================

```

--- a/tests/test_pytest_plugin.py
+++ b/tests/test_pytest_plugin.py
@@ -61,7 +61,7 @@ def test_plugin(testdir: Testdir) -> None:
         """
     )
 
-    result = testdir.runpytest('-v')
+    result = testdir.runpytest('-v', '-p', 'no:asyncio')
     result.assert_outcomes(passed=3 * len(get_all_backends()), skipped=len(get_all_backends()))
 
 
@@ -134,7 +134,7 @@ def test_asyncio(testdir: Testdir) -> None:
         """
     )
 
-    result = testdir.runpytest('-v')
+    result = testdir.runpytest('-v', '-p', 'no:asyncio')
     result.assert_outcomes(passed=2, failed=1, errors=2)
 
 
@@ -171,7 +171,7 @@ def test_autouse_async_fixture(testdir: Testdir) -> None:
         """
     )
 
-    result = testdir.runpytest('-v')
+    result = testdir.runpytest('-v', '-p', 'no:asyncio')
     result.assert_outcomes(passed=len(get_all_backends()))
 
 
@@ -198,7 +198,7 @@ def test_cancel_scope_in_asyncgen_fixture(testdir: Testdir) -> None:
         """
     )
 
-    result = testdir.runpytest('-v')
+    result = testdir.runpytest('-v', '-p', 'no:asyncio')
     result.assert_outcomes(passed=len(get_all_backends()))
 
 
@@ -229,7 +229,7 @@ def test_hypothesis_module_mark(testdir: Testdir) -> None:
         """
     )
 
-    result = testdir.runpytest('-v')
+    result = testdir.runpytest('-v', '-p', 'no:asyncio')
     result.assert_outcomes(passed=len(get_all_backends()) + 1, xfailed=len(get_all_backends()))
 
 
@@ -268,5 +268,5 @@ def test_hypothesis_function_mark(testdir: Testdir) -> None:
         """
     )
 
-    result = testdir.runpytest('-v')
+    result = testdir.runpytest('-v', '-p', 'no:asyncio')
     result.assert_outcomes(passed=2 * len(get_all_backends()), xfailed=2 * len(get_all_backends()))

